client-device-connection-infinispan/src/main/java/org/eclipse/hono/deviceconnection/infinispan/client/HotrodCache.java

```diff
@@ -0,0 +1,279 @@
+/**
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+
+package org.eclipse.hono.deviceconnection.infinispan.client;
+
+import java.net.HttpURLConnection;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.eclipse.hono.client.ConnectionLifecycle;
+import org.eclipse.hono.client.DisconnectListener;
+import org.eclipse.hono.client.ReconnectListener;
+import org.eclipse.hono.client.ServerErrorException;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.commons.api.BasicCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A HotrodCache.
+ *
+ * @param <K> The type of keys used by the cache.
+ * @param <V> The type of values stored in the cache.
+ */
+public final class HotrodCache<K, V> implements ConnectionLifecycle<HotrodCache<K, V>> {
```

Comment:
I find having another cache abstraction on top of the existing one a bit of over-complication. We have an abstraction in enmasse of CacheProvider, which adds necessary cache handling but then services use infinispan interfaces directly. 

https://github.com/EnMasseProject/enmasse/blob/master/iot/iot-infinispan-base/src/main/java/io/enmasse/iot/infinispan/cache/AbstractCacheProvider.java

Maybe at some point we can align two implementations more.

Reply:
The HotrodCache is adding the connection checking capability that we need for the readiness and liveness checks. Apart from that, it doesn't have much value. But we do need these checks in multiple places so I thought it would be a good idea to have them implemented in a single place.

Reply:
Yeah, we created some separate components for infinispan registry that does the same. The only thing here is that this abstraction now "narrows" the original cache api, which the introduce one more thing to maintain going forward.
Maybe another topic for F2F at BCX to see if we can get two implementations more inlined. And maintained at one place.